hipcheck/src/policy_exprs/expr.rs

```diff
@@ -132,6 +218,177 @@ impl From<Primitive> for Expr {
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FunctionType {
+	pub def: FunctionDef,
+	pub arg_tys: Vec<Type>,
+}
+impl FunctionType {
+	pub fn get_return_type(&self) -> Result<ReturnableType> {
+		self.def.type_check(&self.arg_tys)
+	}
+}
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Type {
+	Primitive(PrimitiveType),
+	Function(FunctionType),
+	Lambda(FunctionType),
+	Array(ArrayType),
+	Unknown,
+}
+impl Type {
+	pub fn get_return_type(&self) -> Result<ReturnableType> {
+		self.try_into()
+	}
+}
+impl TryFrom<&Type> for ReturnableType {
+	type Error = crate::policy_exprs::Error;
+	fn try_from(value: &Type) -> Result<ReturnableType> {
+		Ok(match value {
+			Type::Function(fn_ty) | Type::Lambda(fn_ty) => fn_ty.get_return_type()?,
+			Type::Array(arr_ty) => ReturnableType::Array(*arr_ty),
+			Type::Primitive(PrimitiveType::Ident) => ReturnableType::Unknown,
+			Type::Primitive(p_ty) => ReturnableType::Primitive(*p_ty),
+			Type::Unknown => ReturnableType::Unknown,
+		})
+	}
+}
+impl Display for PrimitiveType {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+		write!(f, "{:?}", self)
+	}
+}
+// impl Display for ArrayType {
+// 	fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+// 		write!(f, "{:?}", self)
+// 	}
+// }
+
+pub trait Typed {
+	fn get_type(&self) -> Result<Type>;
+}
+impl Typed for Primitive {
+	fn get_type(&self) -> Result<Type> {
+		Ok(Type::Primitive(self.get_primitive_type()))
+	}
+}
+impl Typed for Array {
+	fn get_type(&self) -> Result<Type> {
+		let mut ty: Option<PrimitiveType> = None;
+		for (idx, elt) in self.elts.iter().enumerate() {
+			let curr_ty = elt.get_primitive_type();
+			if let Some(expected_ty) = ty {
+				if expected_ty != curr_ty {
+					return Err(Error::BadArrayElt {
+						idx,
+						expected: expected_ty,
+						got: curr_ty,
+					});
+				}
+			} else {
+				ty = Some(elt.get_primitive_type());
+			}
+		}
+		Ok(Type::Array(ty))
+	}
+}
+impl Typed for Function {
+	fn get_type(&self) -> Result<Type> {
+		use FuncReturnType::*;
+		let Some(def) = self.opt_def.clone() else {
+			return Err(Error::UnknownFunction(self.ident.0.clone()));
+		};
+		let arg_tys: Vec<Type> = self
+			.args
+			.iter()
+			.map(Typed::get_type)
+			.collect::<Result<Vec<_>>>()?;
+		let fn_type = FunctionType { def, arg_tys };
+		if fn_type.arg_tys.len() == fn_type.def.expected_args - 1 {
+			Ok(Type::Lambda(fn_type))
+		} else {
+			Ok(fn_type.into())
+		}
+	}
+}
+
+impl Typed for Lambda {
+	// @Todo - Lambda should be a FunctionType that takes 1 argument and
+	// contains an interior reference to the function it wraps.
+	// To get its return type, we should combine Unknown with the
+	// other typed args to the function and evaluate.
+	fn get_type(&self) -> Result<Type> {
+		let Expr::Function(func) = self.body.as_ref() else {
+			return Err(Error::BadType("Expected lambda to contain a func"));
+		};
+		let Type::Function(fty) = func.get_type()? else {
+			unreachable!()
```

Comment:
Is this truly an unreachable condition, or does `unreachable!()` here just serve as a placeholder for unimplemented functionality?

If it's really unreachable: It might be better to return an internal error here, instead of panicking. And including comments on why this should be impossible.

If it's a placeholder: It would be clearer to use `todo!()` instead.